add: new section Kotlin compiler fallback strategy (#3340)
JetBrains/kotlin-web-site: docs/topics/gradle/gradle-compilation-and-caches.md

</tab>
</tabs>

## Kotlin compiler fallback strategy

The Kotlin compiler's fallback strategy is to run a compilation outside a Kotlin daemon if the daemon somehow fails. 
If the Gradle daemon is on, the compiler uses the ["In process" strategy](#defining-kotlin-compiler-execution-strategy). 
If the Gradle daemon is off, the compiler uses the "Out of process" strategy.

When this fallback happens, you have the following warning lines in your Gradle's build output:

```
Failed to compile with Kotlin daemon: java.lang.RuntimeException: Could not connect to Kotlin compile daemon
[exception stacktrace]
Using fallback strategy: Compile without Kotlin daemon
Try ./gradlew --stop if this issue persists.
```

However, a silent fallback to another strategy can consume a lot of system resources or lead to non-deterministic builds, 
read more about this in this [YouTrack issue](https://youtrack.jetbrains.com/issue/KT-48843/Add-ability-to-disable-Kotlin-daemon-fallback-strategy).

> To avoid this, there is a Gradle property `kotlin.daemon.useFallbackStrategy`, whose default value is `true`. 

Author: @sarahhaggarty Could you please help me with this sentence? Does it sound weird with this beginning?

Author: Thanks!

</tab>
</tabs>

## Kotlin compiler fallback strategy

The Kotlin compiler's fallback strategy is to run a compilation outside a Kotlin daemon if the daemon somehow fails. 
If the Gradle daemon is on, the compiler uses the ["In process" strategy](#defining-kotlin-compiler-execution-strategy). 
If the Gradle daemon is off, the compiler uses the "Out of process" strategy.

When this fallback happens, you have the following warning lines in your Gradle's build output:

```

> Could not perform incremental compilation: Could not connect to Kotlin compile daemon
> Could not connect to kotlin daemon. Using fallback strategy.

Reviewer: Since 1.8.0 the message is slightly changed:
```suggestion
Failed to compile with Kotlin daemon: java.lang.RuntimeException: Could not connect to Kotlin compile daemon
[exception stacktrace]
Using fallback strategy: Compile without Kotlin daemon
Try ./gradlew --stop if this issue persists.
```

Reviewer: In the case of disabled fallback strategy the message would look like:
```
Failed to compile with Kotlin daemon. Fallback strategy (compiling without Kotlin daemon) is turned off. Try ./gradlew --stop if this issue persists.
```

Author: Thanks! Seems like this message 
>In the case of disabled fallback strategy the message would look like:

is not needed in docs, right?

Reviewer: I think yes, it's not necessary here

```

However, a silent fallback to another strategy can consume a lot of system resources or lead to non-deterministic builds, 
read more about this in this [YouTrack issue](https://youtrack.jetbrains.com/issue/KT-48843/Add-ability-to-disable-Kotlin-daemon-fallback-strategy).
To avoid this, there is a Gradle property `kotlin.daemon.useFallbackStrategy`, whose default value is `true`. 
When the value is `false`, builds fail on problems with the daemon's startup or communication. Declare this property in
`gradle.properties`:

```properties
kotlin.daemon.useFallbackStrategy=false
```

There is also a `useDaemonFallbackStrategy` property in Kotlin compile tasks, which takes priority over the Gradle property if you use both. 

<tabs group="build-script">
<tab title="Kotlin" group-key="kotlin">

```kotlin

> import org.jetbrains.kotlin.gradle.tasks.KotlinCompile
> 
> val compileKotlin: KotlinCompile by tasks
> 
> compileKotlin {
>     useDaemonFallbackStrategy = false
> }

Reviewer: Such syntax with assigning won't work in Kotlin DSL until https://github.com/gradle/gradle/issues/9268 is resolved. Gradle is actively working on it now.

Also the example promotes non-lazy way to configure tasks (non-conformant with [Task Configuration Avoidance](https://docs.gradle.org/current/userguide/task_configuration_avoidance.html)).
```suggestion
tasks {
    compileKotlin {
        useDaemonFallbackStrategy.set(false)
    }   
}
```

```

</tab>
<tab title="Groovy" group-key="groovy">

```groovy

> import org.jetbrains.kotlin.gradle.tasks.KotlinCompile
> 
> val compileKotlin: KotlinCompile by tasks
> 
> compileKotlin {
>     useDaemonFallbackStrategy = false
> }

Reviewer: It's also Kotlin DSL 😅 
I've suggested a Groovy variant also respecting the Task Configuration Avoidance feature
```suggestion
tasks.named("compileKotlin").configure {
    useDaemonFallbackStrategy = false
}
```
